[GSoC_M1.6]: Add a seperate section in translation review modal to show images with their alt text images

Ash-2k3 opened the pull request:
## Overview

This PR does the following: This PR introduces a new section in translation-review-modal which shows images with caption and alt texts/descriptions. This way, reviewers won't have to first click on edit and the select the image to see alt text

We are doing this by making a new function in translation-review-modal-component.ts which will generate a html string which consists of images for both content_html string and translated one too.
These strings will be passed to rte-output component which will render the images and the text tags. 

Mock : https://docs.google.com/document/d/18rFozOzbcDYWa6sS9gTYu4R_VfMevbM5v3xlf7Gy3zQ/edit?pli=1#bookmark=id.xari96c5wxef

## Essential Checklist

- [ ] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [ ] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [ ] The linter/Karma presubmit checks have passed on my local machine.
- [ ] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [ ] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).

## Proof that changes are correct
<img width="1440" alt="Screenshot 2023-07-05 at 22 26 10" src="https://github.com/oppia/oppia/assets/98253080/66cd4814-8885-47f9-ba73-a36fdee5e119">

## Proof of changes on mobile phone
<img width="1440" alt="Screenshot 2023-07-05 at 22 26 19" src="https://github.com/oppia/oppia/assets/98253080/83c47015-e3fc-4a06-af5e-bb37866733f1">

## Proof that changes doesn't cause any breakage in existing image component


https://github.com/oppia/oppia/assets/98253080/e825f8f2-d093-41d7-8ad1-dbddb757399b

Ash-2k3 committed:
Add a seperate section in translation review modal to show images

Ash-2k3 committed:
Add coverage tests for image components

Ash-2k3 committed:
Fix failing acceptance tests

Ash-2k3 committed:
Fix designs

Ash-2k3 commented:
@kevintab95 @chris7716 PTAL, Thanks.

chris7716 commented on core/templates/pages/contributor-dashboard-page/modal-templates/translation-suggestion-review-modal.component.ts:
```diff
@@ -545,4 +552,17 @@ export class TranslationSuggestionReviewModalComponent implements OnInit {
+  getImageInfoForSuggestion(content: string | string[]): string {
+    let htmlString = '';
+
+    if (!Array.isArray(content)) {
```
Why do we check for this? Could you please explain it before the condition by a comment?

Ash-2k3 replied:
Images are present in the form of the string not an array. Added a comment too.

chris7716 commented on core/templates/pages/contributor-dashboard-page/modal-templates/translation-suggestion-review-modal.component.ts:
```diff
@@ -545,4 +552,17 @@ export class TranslationSuggestionReviewModalComponent implements OnInit {
+
+    if (!Array.isArray(content)) {
+      this.showAltText = true;
+      const doc = new DOMParser().parseFromString(content, 'text/html');
```
Also here could you please explain what you are doing by a comment?

Ash-2k3 replied:
I have now added a comment to explain what the function does.

chris7716 commented on extensions/rich_text_components/Image/directives/oppia-noninteractive-image.component.spec.ts:
```diff
@@ -323,4 +323,13 @@ describe('NoninteractiveImage', () => {
       // This is tested to make sure the function did no continue to execute.
       expect(imagePreloaderService.getDimensionsOfImage).not.toHaveBeenCalled();
     });
+  it('should show alt text images when showAltText property is true', () => {
```
Can you add a test for `showAltText = false` case?

Ash-2k3 replied:
Done, thanks

chris7716 reviewed (changes requested):
@Ash-2k3 Thanks! Took a pass.

Ash-2k3 committed:
Add function description and comments

Ash-2k3 committed:
Increase test coverage

Ash-2k3 commented:
Thanks @chris7716, I have addressed your comments, PTAL.

Ash-2k3 committed:
Fix failing tests

Ash-2k3 committed:
Fix frontend tests

Ash-2k3 committed:
Fix linting

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

Ash-2k3 committed:
Make frontend changes

Ash-2k3 commented:
@chris7716 I have made the design changes which we discussed in meeting, PTAL.

Ash-2k3 committed:
Fix linting

chris7716 reviewed (changes requested):
@Ash-2k3 Thanks! This looks more promising. But I have one suggestion. Can we have margins inside the boxes as it is in mocks? Also can we make mobile views much similar to the desktop view(i.e. width of the caption box being inline with above boxes and margins etc.).

chris7716 commented:
Also @kevintab95 could you please have a look?

Ash-2k3 committed:
Fix mobile layout

Ash-2k3 commented:
> @Ash-2k3 Thanks! This looks more promising. But I have one suggestion. Can we have margins inside the boxes as it is in mocks? Also can we make mobile views much similar to the desktop view(i.e. width of the caption box being inline with above boxes and margins etc.).

There were no margins in the mocks, images are at the left most section of the container, however to distinguish among different images, I have added margin bottom. Also Fixed the mobile UI, Thanks @chris7716 .

chris7716 reviewed (commented):
![Screenshot 2023-07-03 at 23 10 57](https://github.com/oppia/oppia/assets/41409951/cbf20751-98e5-40bd-ba3f-e0093195e7a8)
![Screenshot 2023-07-03 at 23 11 42](https://github.com/oppia/oppia/assets/41409951/cbfedbbc-d24e-4aee-8510-f7035be58328)


I mean there should be some padding inside the box as above I think.

kevintab95 commented on extensions/rich_text_components/Image/directives/image.component.html:
```diff
@@ -1,11 +1,19 @@
          [ngStyle]="imageContainerStyle">
     <figcaption class="text-center">{{imageCaption}}</figcaption>
   </div>
+  <div *ngIf="!isLoadingIndicatorShown && !isTryAgainShown && showAltText">
+    <img [src]="imageUrl"
+         alt="{{imageAltText}}"
+         class="e2e-test-image mt-5"
+         [ngStyle]="imageContainerStyle">
+    <figcaption><strong>Caption:</strong> {{imageCaption}}</figcaption>
+    <figcaption class="alt-text mb-5"><strong>Description:</strong> {{imageAltText}}</figcaption>
+  </div>
```
Could you check if the image component works the same way it used to in the exploration editor page? Please add a video in the PR description for the same.

Ash-2k3 replied:
Yup, image component works the same for all the other cases. Please have a look at the PR description for the video, thanks.

kevintab95 commented on extensions/rich_text_components/rte-output-display.component.ts:
```diff
@@ -51,6 +51,7 @@ export class RteOutputDisplayComponent implements AfterViewInit {
   @ViewChild('tabs') tabsTagPortal: TemplateRef<unknown>;
   @ViewChild('video') videoTagPortal: TemplateRef<unknown>;
   @Input() rteString: string;
+  @Input() showAltText: boolean = false;
```
Here and elsewhere, avoid starting a variable name with a verb. See [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide#javascript) for suggestions.

Ash-2k3 replied:
Done, PTAL.

Ash-2k3 committed:
Requested changes applied

Ash-2k3 commented:
> ![Screenshot 2023-07-03 at 23 10 57](https://user-images.githubusercontent.com/41409951/250631007-cbf20751-98e5-40bd-ba3f-e0093195e7a8.png) ![Screenshot 2023-07-03 at 23 11 42](https://user-images.githubusercontent.com/41409951/250631015-cbfedbbc-d24e-4aee-8510-f7035be58328.png)
> 
> I mean there should be some padding inside the box as above I think.

Thanks @chris7716, I have added some padding, PTAL.

chris7716 reviewed (approved):
LGTM!

Ash-2k3 committed:
Fix linting

Ash-2k3 committed:
Fix linting

Ash-2k3 commented:
@kevintab95 PTAL.

kevintab95 commented on core/templates/pages/contributor-dashboard-page/modal-templates/translation-suggestion-review-modal.component.html:
```diff
@@ -133,6 +133,24 @@ <h4 class="oppia-translation-modal-author-header">
+      <oppia-rte-output-display *ngIf="!contentTypeIsSetOfStrings"
+                                  [rteString]="explorationImagesString"
+                                  [altTextIsDisplayed]="altTextIsDisplayed"
+                                  #contentPanel>
```
Is it expected that the "oppia-rte-output-display" components displayed have the same template ref i.e. `#contentPanel` (there are 3 in this component with the same template ref ID)?

Ash-2k3 replied:
Well, for this case I think it's fine to have the same template. Since we have included a *ngIf for this template, I think it's suffice as the diff is only a text section which shows alt text. WDYT?

kevintab95 replied:
Sorry, I don't understand "we have included a *ngIf for this template". It seems like the ngIfs are basically the same for all the components with identical refs.

Ash-2k3 replied:
Ah, Sorry for the misunderstanding, I didn't get it in the first read. I have given it a different temp ref id to avoid ambiguity.

kevintab95 reviewed (changes requested):
Thanks @Ash-2k3! Just one comment

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

Ash-2k3 committed:
Have seperate ref id for templates

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

Ash-2k3 committed:
Fix lint

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

kevintab95 reviewed (approved):
LGTM, thanks @Ash-2k3!

Ash-2k3 committed:
Merge branch 'develop' into GSoC_M1.6_Image_Alt_Text

U8NWXD commented:
@Ash-2k3 I think this PR might have introduced a flaky frontend test. Specifically, the [`NoninteractiveImage should not show alt text images when altTextIsDisplayed property isfalse`](https://github.com/oppia/oppia/blame/525bc952f4afe096fc22fe3f3f7e17f6bd194d1f/extensions/rich_text_components/Image/directives/oppia-noninteractive-image.component.spec.ts#L338) test was introduced by this PR and has since failed on two PRs:

* #18704: https://github.com/oppia/oppia/actions/runs/5702183651/job/15453666603#step:9:7470
  * Passed on re-run, suggesting that this is indeed a flake: https://github.com/oppia/oppia/actions/runs/5702183651
* #18706: https://github.com/oppia/oppia/actions/runs/5703520198/job/15456249876?pr=18706#step:9:6331
  * Passed on re-run, suggesting that this is indeed a flake: https://github.com/oppia/oppia/actions/runs/5703520198?pr=18706

Could you please investigate? If the flake starts impeding the developer workflow, we may need to revert this PR.

Ash-2k3 commented:
I see @U8NWXD, I will take a look, Thanks.

U8NWXD commented:
This comment is just to document how we followed the [revert and regression policy](https://github.com/oppia/oppia/wiki/Revert-and-Regression-Policy) here:

When I initially commented, I wasn't sure how much this flake was impeding the developer workflow since it had only happened once (I thought the failure in my revert PR might have been because of some problem with the revert, not a flake in develop) and I hadn't heard from any developers that this flake was blocking them.

When I checked back ~12 hours later, the flake had occurred in more PRs, so it was likely impacting developers. By that time a fix-forward PR was already open, so the fastest path to fix develop was to fix-forward rather than revert.
